src/client/websocket/WebSocketShard.js

```diff
@@ -0,0 +1,362 @@
+     * @type {number}
+     */
+    this.status = Status.IDLE;
+
+    /**
+     * The current sequence of the WebSocket
+     * @type {number}
+     */
+    this.sequence = -1;
+    this.pings = [];
+    this.lastPingTimestamp = -1;
+    this.sessionID = undefined;
+
+    this.trace = undefined;
+
+    /**
+     * Contains the rate limit queue and metadata
+     * @type {Object}
+     */
+    this.ratelimit = {
+      queue: [],
+      total: 120,
+      remaining: 120,
+      time: 60e3,
+      timer: null,
+    };
+
+    this.inflate = null;
+
+    this.connect();
+  }
+
+  /**
+   * Average heartbeat ping of the websocket, obtained by averaging the {@link Client#pings} property
+   * @type {number}
+   * @readonly
+   */
+  get ping() {
+    const sum = this.pings.reduce((a, b) => a + b, 0);
+    return sum / this.pings.length;
+  }
+
+  /**
+   * Emits a debug event.
+   * @param {string} message Debug message
+   */
+  debug(message) {
+    this.manager.debug(`[shard ${this.id}] ${message}`);
+  }
+
+  /**
+   * Sends a heartbeat or sets an interval for sending heartbeats.
+   * @param {number} [time] If -1, clears the interval, any other number sets an interval
+   * If no value is given, a heartbeat will be sent instantly
+   */
+  heartbeat(time) {
+    if (!isNaN(time)) {
+      if (time === -1) {
+        this.debug('Clearing heartbeat interval');
+        this.manager.client.clearInterval(this.heartbeatInterval);
+        this.heartbeatInterval = null;
+      } else {
+        this.debug(`Setting a heartbeat interval for ${time}ms`);
+        this.heartbeatInterval = this.manager.client.setInterval(() => this.heartbeat(), time);
+      }
+      return;
+    }
+
+    this.debug('Sending a heartbeat');
+    this.lastPingTimestamp = Date.now();
+    this.send({
+      op: OPCodes.HEARTBEAT,
+      d: this.sequence,
+    });
+  }
+
+  /**
+   * Acknowledges a heartbeat.
+   */
+  ackHeartbeat() {
+    const latency = Date.now() - this.lastPingTimestamp;
+    this.debug(`Heartbeat acknowledged, latency of ${latency}ms`);
+    this.pings.unshift(latency);
+    if (this.pings.length > 3) this.pings.length = 3;
+  }
+
+  /**
+   * Connects the shard to a gateway.
+   */
+  connect() {
+    this.inflate = new zlib.Inflate({
+      chunkSize: 65535,
+      flush: zlib.Z_SYNC_FLUSH,
+      to: WebSocket.encoding === 'json' ? 'string' : '',
+    });
+    const gateway = this.manager.gateway;
+    this.debug(`Connecting to ${gateway}`);
+    const ws = this.ws = WebSocket.create(gateway, {
+      v: this.manager.client.options.ws.version,
+      compress: 'zlib-stream',
+    });
+    ws.onopen = this.onOpen.bind(this);
+    ws.onmessage = this.onMessage.bind(this);
+    ws.onerror = this.onError.bind(this);
+    ws.onclose = this.onClose.bind(this);
+    this.status = Status.CONNECTING;
+  }
+
+  /**
+   * Called whenever a packet is received
+   * @param {Object} packet Packet received
+   * @returns {boolean}
+   */
+  onPacket(packet) {
```

Comment:
What did happen to the raw event?

Reply:
it run away but i find it and bring it back 🎉